### ISSUE TITLE
LogColorMapper maps to wrong values

### ISSUE THREAD

**Issue by poplarShift** [bokeh/bokeh#8724]
Possibly related: #8061
Definitely related: #5020

TL;DR I'm happy to make a PR to use `log` for the `LogColorMapper` instead of `log1p` but wanted to hear opinions first.

As of now, LogTicker and LogTickFormatter seem to work with a simple log(x) transform, whereas LogColorMapper does a log(1+x) transform (`log1p`), which makes the colorbar map to the wrong values. The documentation does not mention the occurence of `log1p` at all.

So one of these two options has to be used for both cases.

Scientifically, `log1p` makes absolutely no sense because it makes the transform depend on the unit of the measurement, but in #5020 the reason given was that small values smaller than 10-ish are "usually below the range of values that log is used for", which I don't understand at all.

I get the concern that for 0-values, the log scale will blow up, but that's a problem of choosing the right transform for your data and should be handled by the user, not by bokeh. One could always add separate `Log1pColorMapper`, `Log1pTickFormatter`, etc. or something like it if users really need this behaviour for their dimensionless values.

**Comment by philippjfr** [comment]
Unfortunately at this point I think maintaining backwards compatibility for `LogColorMapper` would be necessary, so I'd suggest that we could simply add an option to switch between ``log``, ``log1p`` and ``log10`` and make it very clear in the docs that the default is ``log1p``. The default could then be reconsidered for bokeh 2.0.

**Comment by bryevdv** [comment]
> I get the concern that for 0-values, the log scale will blow up, but that's a problem of choosing the right transform for your data and should be handled by the user, not by bokeh.

This is easy to say when you aren't on the receiving end of support requests and bug reports. :) Users expect these things to "just work" and complain when they don't. Certainly other libraries, e.g. MPL make log scale usage "including zero" work in various ways, we just have not had the resources to devote to doing the same as of yet. 

> Unfortunately at this point I think maintaining backwards compatibility for LogColorMapper would be necessary,

I could be swayed that the current implementation is approximately useful/correct for "large N" but broken (i.e. in need of a *fix*) for "small N", and that therefore solution that works for both regimes by default is permissible.

**Comment by philippjfr** [comment]
Fair enough, I agree that this could be considered broken particularly since none of the docs ever mentioned the fact that it was actually using log1p.

**Comment by jbednar** [comment]
>  in #5020 the reason given was that small values smaller than 10-ish are "usually below the range of values that log is used for", which I don't understand at all.

I think that's a meaningful claim, but only in the context of whole numbers (e.g. counts).  For counts, one would be unlikely to select a log scale unless there are some fairly large counts, because if the largest count is under 100 or so all the values should be visible and distinguishable even in a linear scale. And counts are where zero values are likely to come up, e.g. for a 2D histogram viewed as an image.  But reasoning like that doesn't apply to floats, where "1" can have any arbitrary relationship with the range of data being observed.  So one could imagine using log1p only for integer data types?

Also note that histograms with a log y axis would presumably have similar issues, i.e. could use log1p meaningfully if they are integers, and otherwise log1p would be arbitrary and sometimes completely inappropriate.

**Comment by poplarShift** [comment]
Thanks for the comments. I'll put something together, pretty busy with a conference right now but the way I understand you I will 1) fix the wrong colormap as it is a bug, which is straightforward, and 2) in the mid-term we should figure out a more suitable handling of log-scales in general.

**Comment by bryevdv** [comment]
@poplarShift I can't actually find any evidence that other libraries (Altair/Vegalite, MPL, Plotly) handle zero on log scales, except for MPL's `symlog` which I think is generally regarded as bad. So I am inclined now to agree with you that we should just do the simple thing and stop using `log1p` in the colorbar case, and make it match what the other log scales on axes do. Since that is technically a "breaking change" I will move this to the 2.0 milestone. Just FYI most of the work for 2.0 is already done on the landing-2.0 branch so it should be out a matter of weeks after 1.4 is. 

Thanks for your patience with this old issue.

**Comment by poplarShift** [comment]
Thanks for picking this up again! Are you saying that #8832 should be re-opened?

**Comment by bryevdv** [comment]
@poplarShift If you would like to re-submit that PR against the `landing-2.0` branch it would certainly be helpful (or  alternatively it could be submitted agains `master` after 1.4 is released)
